### PR TITLE
[BEAM-2166] Split Coder's encode/decode methods into two methods depending on context.

### PR DESCRIPTION
**Pull request by robertwb**
…text.

This allows the outer context to be marked deprecated.  A follow-up PR will
remove the old method once all consumers have been updated.

Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`.
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @lukecwik

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StructuredCoder.java [hunk 1]
```diff
@@ -98,6 +101,34 @@ public String toString() {
     return builder.toString();
   }
 
+  public void encode(T value, OutputStream outStream)
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StructuredCoder.java]: Add default implementations of:
* encode(T, OutputStream, Context) that calls the appropriate encode(T, OutputStream)/encodeOuter(T, OutputStream) methods.
* decode(InputStream, Context) that calls the appropriate decode(InputStream)/decodeOuter(InputStream) methods.

This way we can remove the base method as soon as everyone has migrated to implementing the non Context version.

Also, we should have:
* encodeOuter(T, OutputStream) call encode(T, OutputStream)
* decodeOuter(InputStream) call decode(InputStream)

This way only Coders where outer context is important will need to implement it.

Comment (robertwb) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/StructuredCoder.java]: Implemented the base encode/decode methods (though this opens us up to infinite recursion and no abstract methods, it should be short term). However, it's unsafe to unconditionally delegate the Outer ones to nested until everything is migrated.

**Comment by dhalperi** [comment]
There should probably be a JIRA.

Do the other methods that take context (e.g., estimating byte size) need to be updated too?

**Comment by robertwb** [comment]
JIRA filed. Added non-context versions of size estimation. Didn't add an Outer version, as these are estimations and context doesn't ever seem to be actually inspected for nearly any leaf Coders.

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
LGTM

**Pull request closed without merging**